sentinel/src/schedule/outbound.js

```diff
@@ -0,0 +1,259 @@
+  // resulting destination, and the value is either an object path string defining the source path, or
+  // an object with more in-depth config on how to map the value.
+  Object.keys(config.mapping).forEach(dest => {
+    const {
+      path, expr, required
+    } = normaliseSourceConfiguration(config.mapping[dest]);
+
+    let srcValue;
+
+    if (expr) {
+      try {
+        srcValue = vm.runInNewContext(expr, {doc});
+      } catch (err) {
+        throw Error(`Mapping error for '${config.key}/${dest}' JS error on source document: ${doc._id}: ${err}`);
+      }
+    } else {
+      srcValue = objectPath.get({doc}, path);
+    }
+
+    if (required && srcValue === undefined) {
+      throw Error(`Mapping error for '${config.key}/${dest}': cannot find '${path}' on source document`);
+    }
+
+    if (srcValue) {
+      objectPath.set(toReturn, dest, srcValue);
+    }
+  });
+
+  return toReturn;
+};
+
+// Attempts to send a given payload using a given push config
+const send = (payload, config) => {
+  const sendOptions = {
+    method: 'POST',
+    url: urlJoin(config.destination.base_url, config.destination.path),
+    body: payload,
+    json: true
+  };
+
+  const auth = () => {
+    const authConf = config.destination.auth;
+
+    if (!authConf) {
+      return Promise.resolve();
+    }
+
+    if (!authConf.type) {
+      return Promise.reject(Error('No auth.type, either declare the type or omit the auth property'));
+    }
+
+    if (authConf.type.toLowerCase() === 'basic') {
+      return fetchPassword(authConf['password_key'])
+        .then(password => {
+          sendOptions.auth = {
+            username: authConf.username,
+            password: password,
+            sendImmediately: true
+          };
+        });
+    }
+
+    if (authConf.type.toLowerCase() === 'muso-sih') {
+      return fetchPassword(authConf['password_key'])
+        .then(password => {
+          const authOptions = {
+            method: 'POST',
+            form: {
+              login: authConf.username,
+              password: password
+            },
+            url: urlJoin(config.destination.base_url, authConf.path),
+            json: true
+          };
+
+          return request(authOptions)
+            .then(result => {
+              // No that's not a spelling mistake, this API is sometimes French!
+              if (result.statut !== 200) {
+                logger.error('Non-200 status from Muso auth', result);
+                throw new Error(`Got ${result.statut} when requesting auth`);
+              }
+
+              sendOptions.qs = {
+                token: result.data.username_token
+              };
+            });
+        });
+    }
+
+    // Misconfigured auth type
+    return Promise.reject(new Error(`Invalid auth type '${authConf.type}'. Supported: basic, muso-sih`));
+  };
+
+  return auth().then(() => request(sendOptions));
```

Comment:
Minor point - I think this would be slightly cleaner if `auth` returned the options, instead of mutating a shared object like this.

Reply:
Hmmm, I tried this and it read worse to me unfortunately. IDK, something about it being scoped to a reasonably small function anyway makes passing it around feel more confusing than not.